packages/aws-cdk-lib/core/lib/arn.ts

```diff
@@ -268,21 +334,50 @@ export class Arn {
       detectedArnFormat = ArnFormat.NO_RESOURCE_NAME;
     }
 
-    if (slashIndex !== -1) {
-      resource = resourceTypeOrName.substring(resourcePartStartIndex, slashIndex);
-      resourceName = resourceTypeOrName.substring(slashIndex + 1);
-    } else {
-      resource = resourceTypeOrName;
-    }
+    if (arnFormat === ArnFormat.HIERARCHICAL_SLASH_RESOURCE_NAME && options?.hierarchicalDepth) {
+      const resourceParts = resourceTypeOrName.split('/');
+      const targetIndex = (options.hierarchicalDepth - 1) * 2;
+
+      if (targetIndex + 1 >= resourceParts.length) {
+        throw new UnscopedValidationError(`Insufficient hierarchy depth in ARN. Requested depth ${options.hierarchicalDepth} but ARN only has ${Math.floor(resourceParts.length / 2)} levels: ${arn}`);
+      }
 
-    if (rest.length > 0) {
-      if (!resourceName) {
-        resourceName = '';
+      resource = resourceParts[targetIndex];
+      resourceName = resourceParts[targetIndex + 1];
```

Comment:
I am reaching out to some experts to see if we need to support ARNs with something like `type1/resourceId1/type2/resourceId2/with/more/stuff`